packages/sdk/src/filter_next/subscription.ts

```diff
@@ -0,0 +1,589 @@
+    if (this.isMessageReceived(message)) {
+      log.info(
+        `Skipping invoking callbacks for already received message: pubsubTopic:${this.pubsubTopic}, peerId:${_peerId.toString()}, contentTopic:${message.contentTopic}`
+      );
+      return;
+    }
+
+    log.info(`Invoking message for contentTopic: ${message.contentTopic}`);
+
+    this.messageEmitter.dispatchEvent(
+      new CustomEvent<WakuMessage>(message.contentTopic, {
+        detail: message
+      })
+    );
+  }
+
+  private addSingle<T extends IDecodedMessage>(
+    decoder: IDecoder<T>,
+    callback: Callback<T>
+  ): void {
+    log.info(`Adding subscription for contentTopic: ${decoder.contentTopic}`);
+
+    const isNewContentTopic = !this.contentTopics.includes(
+      decoder.contentTopic
+    );
+
+    if (isNewContentTopic) {
+      this.toSubscribeContentTopics.add(decoder.contentTopic);
+    }
+
+    if (this.callbacks.has(decoder)) {
+      log.warn(
+        `Replacing callback associated associated with decoder with pubsubTopic:${decoder.pubsubTopic} and contentTopic:${decoder.contentTopic}`
+      );
+
+      const callback = this.callbacks.get(decoder);
+      this.callbacks.delete(decoder);
+      this.messageEmitter.removeEventListener(decoder.contentTopic, callback);
+    }
+
+    const eventHandler = (event: CustomEvent<WakuMessage>): void => {
+      void (async (): Promise<void> => {
+        try {
+          const message = await decoder.fromProtoObj(
+            decoder.pubsubTopic,
+            event.detail as IProtoMessage
+          );
+          void callback(message!);
+        } catch (err) {
+          log.error("Error decoding message", err);
+        }
+      })();
+    };
+
+    this.callbacks.set(decoder, eventHandler);
+    this.messageEmitter.addEventListener(decoder.contentTopic, eventHandler);
+
+    log.info(
+      `Subscription added for contentTopic: ${decoder.contentTopic}, isNewContentTopic: ${isNewContentTopic}`
+    );
+  }
+
+  private removeSingle<T extends IDecodedMessage>(decoder: IDecoder<T>): void {
+    log.info(`Removing subscription for contentTopic: ${decoder.contentTopic}`);
+
+    const callback = this.callbacks.get(decoder);
+
+    if (!callback) {
+      log.warn(
+        `No callback associated with decoder with pubsubTopic:${decoder.pubsubTopic} and contentTopic:${decoder.contentTopic}`
+      );
+    }
+
+    this.callbacks.delete(decoder);
+    this.messageEmitter.removeEventListener(decoder.contentTopic, callback);
+
+    const isCompletelyRemoved = !this.contentTopics.includes(
+      decoder.contentTopic
+    );
+
+    if (isCompletelyRemoved) {
+      this.toUnsubscribeContentTopics.add(decoder.contentTopic);
+    }
+
+    log.info(
+      `Subscription removed for contentTopic: ${decoder.contentTopic}, isCompletelyRemoved: ${isCompletelyRemoved}`
+    );
+  }
+
+  private isMessageReceived(message: WakuMessage): boolean {
+    try {
+      const messageHash = messageHashStr(
+        this.pubsubTopic,
```

Comment:
I know it's _probably_ unlikely, but this line makes me a little uncomfortable. Are we sure it's impossible for a message that doesn't match the pubsub topic to reach here?

Reply:
> I know it's _probably_ unlikely, but this line makes me a little uncomfortable. Are we sure it's impossible for a message that doesn't match the pubsub topic to get passed as an arg to this function?

Reply:
no, good point, I will add another check tho it shouldn't happen since `filter.ts` ensures it